### PR TITLE
Fix an InvalidCastException in IOperation factory when a name of a range variable conflicts with another range variable.

### LINKED ISSUE #23223: An InvalidCastException is thrown when a name of range variable conflicts with another range variable
```
        <CompilerTrait(CompilerFeature.IOperation)>
        Public Sub DuplicateRangeVariableName_IOperation()
            Dim source = <![CDATA[
Option Strict Off
Imports System
Imports System.Linq

Module Module1
    Sub Main()
        Dim q As Object = From implicit In New Integer() {1, 2, 3} Let implicit = "1" Select implicit 'BIND:"From implicit In New Integer() {1, 2, 3} Let implicit = "1" Select implicit"
    End Sub
End Module]]>.Value

            Dim expectedOperationTree = <![CDATA[
]]>.Value

            Dim expectedDiagnostics = <![CDATA[
BC30978: Range variable 'implicit' hides a variable in an enclosing block or a range variable previously defined in the query expression.
        Dim q As Object = From implicit In New Integer() {1, 2, 3} Let implicit = "1" Select implicit 'BIND:"From implicit In New Integer() {1, 2, 3} Let implicit = "1" Select implicit"
                                                                       ~~~~~~~~
]]>.Value

            VerifyOperationTreeAndDiagnosticsForTest(Of QueryExpressionSyntax)(source, expectedOperationTree, expectedDiagnostics)
        End Sub
```

Observed:
```
Test 'M:Microsoft.CodeAnalysis.VisualBasic.UnitTests.Semantics.QueryExpressions.DiplicateRangeVariableName_IOperation' failed:
	Unable to cast object of type 'Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE.PENamedTypeSymbolWithEmittedNamespaceName' to type 'AnonymousTypePublicSymbol'.
	System.InvalidCastException: Unable to cast object of type 'Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE.PENamedTypeSymbolWithEmittedNamespaceName' to type 'AnonymousTypePublicSymbol'.
	at Microsoft.CodeAnalysis.VisualBasic.LocalRewriter.PopulateRangeVariableMapForAnonymousType(SyntaxNode syntax, BoundExpression anonymousTypeInstance, ImmutableArray`1 rangeVariables, Int32& firstUnmappedRangeVariable, Dictionary`2 rangeVariableMap, Boolean inExpressionLambda)
	at Microsoft.CodeAnalysis.VisualBasic.LocalRewriter.PopulateRangeVariableMapForAnonymousType(SyntaxNode syntax, BoundExpression anonymousTypeInstance, ImmutableArray`1 rangeVariables, Int32& firstUnmappedRangeVariable, Dictionary`2 rangeVariableMap, Boolean inExpressionLambda)
	at Microsoft.CodeAnalysis.VisualBasic.LocalRewriter.PopulateRangeVariableMapForQueryLambdaRewrite(BoundQueryLambda node, Dictionary`2& rangeVariableMap, Boolean inExpressionLambda)
	at Microsoft.CodeAnalysis.Operations.VisualBasicOperationFactory.QueryLambdaRewriterPass1.VisitQueryLambda(BoundQueryLambda node)
	at Microsoft.CodeAnalysis.Operations.VisualBasicOperationFactory.RewriteQueryLambda(BoundQueryLambda node)
	at Microsoft.CodeAnalysis.Operations.VisualBasicOperationFactory.CreateInternal(BoundNode boundNode)
	at Microsoft.CodeAnalysis.Operations.VisualBasicOperationFactory._Lambda$__3-0(BoundNode n)
	at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
	at Microsoft.CodeAnalysis.Operations.VisualBasicOperationFactory.Create(BoundNode boundNode)
	at Microsoft.CodeAnalysis.Operations.VisualBasicOperationFactory._Closure$__121-0._Lambda$__1()
	at System.Lazy`1.CreateValue()
	at System.Lazy`1.LazyInitValue()
	at System.Lazy`1.get_Value()
	at Microsoft.CodeAnalysis.Operations.LazyDelegateCreationExpression.get_TargetImpl()
	at Microsoft.CodeAnalysis.Operations.BaseDelegateCreationExpression.get_Target()
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.VisitDelegateCreation(IDelegateCreationOperation operation)
	at Microsoft.CodeAnalysis.Operations.BaseDelegateCreationExpression.Accept(OperationVisitor visitor)
	at Microsoft.CodeAnalysis.Operations.OperationWalker.Visit(IOperation operation)
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.Visit(IOperation operation)
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.VisitArgument(IArgumentOperation operation)
	at Microsoft.CodeAnalysis.Operations.BaseArgument.Accept(OperationVisitor visitor)
	at Microsoft.CodeAnalysis.Operations.OperationWalker.Visit(IOperation operation)
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.Visit(IOperation operation)
	at Microsoft.CodeAnalysis.Operations.OperationWalker.VisitOperationArrayElement[T](T operation)
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.VisitArrayCommon[T](ImmutableArray`1 list, String header, Boolean logElementCount, Boolean logNullForDefault, Action`1 arrayElementVisitor)
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.VisitArray[T](ImmutableArray`1 list, String header, Boolean logElementCount, Boolean logNullForDefault)
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.VisitArguments(ImmutableArray`1 arguments)
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.VisitInvocation(IInvocationOperation operation)
	at Microsoft.CodeAnalysis.Operations.BaseInvocationExpression.Accept(OperationVisitor visitor)
	at Microsoft.CodeAnalysis.Operations.OperationWalker.Visit(IOperation operation)
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.Visit(IOperation operation)
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.Visit(IOperation operation, String header)
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.VisitTranslatedQuery(ITranslatedQueryOperation operation)
	at Microsoft.CodeAnalysis.Operations.BaseTranslatedQueryExpression.Accept(OperationVisitor visitor)
	at Microsoft.CodeAnalysis.Operations.OperationWalker.Visit(IOperation operation)
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.Visit(IOperation operation)
	at Microsoft.CodeAnalysis.Test.Utilities.OperationTreeVerifier.GetOperationTree(Compilation compilation, IOperation operation, Int32 initialIndent)
	at Microsoft.CodeAnalysis.VisualBasic.UnitTests.BasicTestBaseBase.GetOperationTreeForTest[TSyntaxNode](VisualBasicCompilation compilation, String fileName, Int32 which)
	at Microsoft.CodeAnalysis.VisualBasic.UnitTests.BasicTestBaseBase.VerifyOperationTreeForTest[TSyntaxNode](VisualBasicCompilation compilation, String fileName, String expectedOperationTree, Int32 which, Action`3 additionalOperationTreeVerifier)
	at Microsoft.CodeAnalysis.VisualBasic.UnitTests.BasicTestBaseBase.VerifyOperationTreeAndDiagnosticsForTest[TSyntaxNode](VisualBasicCompilation compilation, String fileName, String expectedOperationTree, String expectedDiagnostics, Int32 which, Action`3 additionalOperationTreeVerifier)
	at Microsoft.CodeAnalysis.VisualBasic.UnitTests.BasicTestBaseBase.VerifyOperationTreeAndDiagnosticsForTest[TSyntaxNode](String testSrc, String expectedOperationTree, String expectedDiagnostics, VisualBasicCompilationOptions compilationOptions, VisualBasicParseOptions parseOptions, Int32 which, IEnumerable`1 additionalReferences, Action`3 additionalOperationTreeVerifier, Boolean useLatestFramework)
	F:\GitHub\roslyn\src\Compilers\VisualBasic\Test\Semantic\Semantics\QueryExpressions.vb(14736,0): at Microsoft.CodeAnalysis.VisualBasic.UnitTests.Semantics.QueryExpressions.DiplicateRangeVariableName_IOperation()
```

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #23223.

**Customer scenario**

Request IOperation tree for a query expression like the one below:
```
Dim q As Object = From implicit In New Integer() {1, 2, 3} Let implicit = "1" Select implicit
```

An InvalidCastException is thrown by IOperation factory.

**Bugs this fixes:**

Fixes #23223.

**Workarounds, if any**

No

**Risk**

Low

**Performance impact**

Low perf impact because no extra allocations/no complexity changes

**Is this a regression from a previous update?**

No

**Root cause analysis:**

A test gap. Unit-tests are added.

**How was the bug found?**

Running existing compiler tests with IOperation validation test hook enabled.

### CODE REVIEW

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/QueryExpressions.vb [hunk 1]
```diff
@@ -14710,6 +14710,77 @@ ITranslatedQueryOperation (OperationKind.TranslatedQuery, Type: QueryAble) (Synt
 
             VerifyOperationTreeAndDiagnosticsForTest(Of QueryExpressionSyntax)(source, expectedOperationTree, expectedDiagnostics)
         End Sub
+
+        <CompilerTrait(CompilerFeature.IOperation)>
+        <WorkItem(23223, "https://github.com/dotnet/roslyn/issues/23223")>
+        Public Sub DiplicateRangeVariableName_IOperation()
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/QueryExpressions.vb]: Typo: `Duplicate`

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review IOperation fix.

**Comment by AlekseyTs** [comment]
@jinujoseph FYI

**Review by mavasani — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_Query.vb [hunk 2]
```diff
@@ -157,9 +157,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
                 Dim propertyDefName As String = propertyDef.Name
 
-                If propertyDefName.StartsWith("$"c, StringComparison.Ordinal) AndAlso Not String.Equals(propertyDefName, StringConstants.Group, StringComparison.Ordinal) Then
+                If propertyDefName.StartsWith("$"c, StringComparison.Ordinal) AndAlso
+                   (propertyDefName.Equals(StringConstants.It) OrElse propertyDefName.Equals(StringConstants.It1) OrElse propertyDefName.Equals(StringConstants.It2)) Then
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_Query.vb]: Consider including an explicit `StringComparison` in `Equals()` calls.

Comment (cston) [reply src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_Query.vb]: How do we hit the three different cases?

**Review by cston — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_Query.vb (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_Query.vb]: > Consider including an explicit StringComparison in Equals() calls.

Added

**Comment by natidea** [comment]
@AlekseyTs please retarget this PR to dev15.5.x following Jason's email this morning

**Comment by AlekseyTs** [comment]
@natidea 
> please retarget this PR to dev15.5.x following Jason's email this morning

Done.

**Comment by natidea** [comment]
Shiproom approved to merge via https://devdiv.visualstudio.com/DevDiv/_workitems/edit/525901

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review IOperation fix, need a second sign-off for this 15.5 fix.

**Comment by AlekseyTs** [comment]
Test windows_debug_vs-integration_prtest please.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review IOperation fix, need a second sign-off for this 15.5 fix.

**Comment by agocke** [comment]
Reviewing

**Review by agocke — APPROVED** [review]
LGTM

**Pull request merged**